File: backend/app/schemas/market_data.py
from decimal import Decimal

from pydantic import BaseModel


class PriceResponse(BaseModel):
    ticker: str
    price: Decimal


class PricePoint(BaseModel):
    timestamp: str
    close: float


class QuoteStats(BaseModel):
    open: float | None = None
    high: float | None = None
    low: float | None = None
    volume: int | None = None
    fifty_two_week_high: float | None = None
    fifty_two_week_low: float | None = None
    avg_volume: int | None = None


class QuoteResponse(BaseModel):
    ticker: str
    name: str | None = None
    exchange: str | None = None
    currency: str | None = None
    price: float
    previous_close: float | None = None
    change: float
    change_percent: float | None = None
    points: list[PricePoint]
    stats: QuoteStats


class NewsItem(BaseModel):
    title: str
    summary: str | None = None
    published_at: str | None = None
    source: str | None = None
    url: str | None = None


class NewsResponse(BaseModel):
    ticker: str
    items: list[NewsItem]

File: frontend/src/types/market-data.ts
/**
 * TypeScript types that mirror the backend Pydantic schemas.
 *
 * Keep these in sync with:
 *   backend/app/schemas/market_data.py
 */

/** Chart lookback windows supported by GET /market-data/quote/:ticker */
export const TIME_RANGES = [
  "1D",
  "1W",
  "1M",
  "3M",
  "6M",
  "YTD",
  "1Y",
  "2Y",
  "5Y",
  "10Y",
  "ALL",
] as const;

export type TimeRange = (typeof TIME_RANGES)[number];

export type PricePoint = {
  timestamp: string;
  close: number;
};

/** Stats grid values. Any field the backend can't source is null → rendered as "–". */
export type QuoteStats = {
  open: number | null;
  high: number | null;
  low: number | null;
  volume: number | null;
  fifty_two_week_high: number | null;
  fifty_two_week_low: number | null;
  avg_volume: number | null;
};

/** Response shape for GET /market-data/quote/:ticker */
export type Quote = {
  ticker: string;
  name: string | null;
  exchange: string | null;
  currency: string | null;
  price: number;
  previous_close: number | null;
  change: number;
  change_percent: number | null;
  points: PricePoint[];
  stats: QuoteStats;
};

/** Single headline from GET /market-data/news/:ticker */
export type NewsItem = {
  title: string;
  summary: string | null;
  published_at: string | null;
  source: string | null;
  url: string | null;
};

/** Response shape for GET /market-data/news/:ticker */
export type NewsResponse = {
  ticker: string;
  items: NewsItem[];
};
